library/src/updater.rs

```diff
@@ -223,11 +221,37 @@ fn update_internal(_: &UpdaterLockState) -> anyhow::Result<UpdateStatus> {
     // Saves state to disk (holds Config lock while writing).
 
     let config = copy_update_config()?;
+    // We should never try to write this state as some other writer may be
+    // racing with us, we should get a new state inside a lock if we want
+    // to write.
+    let read_only_state =
+        UpdaterState::load_or_new_on_error(&config.cache_dir, &config.release_version);
+
+    // We discard any events if we have more than 3 queued to make sure
+    // we don't stall the client.
+    let events = read_only_state.copy_events(3);
+    for event in events {
+        let result = crate::network::send_patch_event(event, &config);
+        if let Err(err) = result {
+            error!("Failed to report event: {:?}", err);
+        }
+    }
+    // We're abusing the config lock as a UpdateState lock for now.
+    let read_only_state = with_config(|_| {
+        let mut state =
+            UpdaterState::load_or_new_on_error(&config.cache_dir, &config.release_version);
+        // This will clear any events which got queued between the time we
+        // loaded the state now, but that's OK for now.
+        let result = state.clear_events();
```

Comment:
How does this change get saved?

Reply:
```    pub fn clear_events(&mut self) -> Result<()> {
        self.queued_events.clear();
        self.save()
    }```

Reply:
Oh interesting. I wonder if the "save at the end of a mutating method" pattern is something we want to either adopt throughout or discourage, as it's currently done inconsistently in `UpdaterState`.

Reply:
Yeah, the inconsistency is the problem.  I suspect what we actually want is some sort of file system object which you have to get a lock to use?  And then UpdaterState doesnt' know how to save itself, rather it gets passed to something that knows how to read or write it?